File: static/typescript/0.1.0/src/fieldExpr.ts
export type Scalar =
  | string
  | number
  | boolean
  | Date;

export type UnaryOperator = 'not' | '-';
export type BooleanOperator =
  | 'eq'
  | 'neq'
  | 'gt'
  | 'gte'
  | 'lt'
  | 'lte'
  | 'and'
  | 'or'
  | 'not'
  | 'like'
  | 'in'
  | 'inPast';

export type ArithmeticOperator = '+' | '-' | '*' | '/';

export type AggregateOperator =
  | 'min'
  | 'max'
  | 'count'
  | 'countDistinct'
  | 'avg'
  | 'avgDistinct';

export type DateOperator = 'day' | 'month' | 'year';
export type TimeOperator = 'hour' | 'minute' | 'second' | 'millisecond';
export type ProjectionOperator = DateOperator | TimeOperator;

export type DateGranularity = 'years' | 'months' | 'weeks' | 'days';
export type TimeGranularity = 'hours' | 'minutes' | 'seconds' | 'millis';
export type DateTimeGranularity = DateGranularity | TimeGranularity;

export type Operator =
  | UnaryOperator
  | BooleanOperator
  | ArithmeticOperator
  | AggregateOperator
  | ProjectionOperator
  | 'ident';

export type Expr = FieldExpr | Scalar;

/**
 *  A tree of expressions, each of which has an associated name.
 */
export abstract class FieldExpr {
  // A human-readable representation of the expression. Use this to refer to the
  // expression in a `select` or `orderBy`.
  name: string;

  // User-specified alias for expression.
  alias?: string;

  constructor(name: string) {
    this.name = name;
  }

  as(alias: string): FieldExpr {
    this.alias = alias;
    return this;
  }

  toString(): string {
    return this.name;
  }

  abstract operator(): Operator;
  abstract operands(): Expr[]
}

export class UnaryFieldExpr extends FieldExpr {
  field: FieldExpr;
  op: UnaryOperator;

  constructor(field: FieldExpr, op: UnaryOperator) {
    super(`(${op}(${field.name}))`);
    this.field = field;
    this.op = op;
  }

  operator(): Operator {
    return this.op;
  }

  operands(): Expr[] {
    return [this.field];
  }
}

export class BooleanFieldExpr extends FieldExpr {
  field: FieldExpr;
  op: BooleanOperator;
  other: FieldExpr;

  constructor(
    field: FieldExpr,
    op: BooleanOperator,
    other: FieldExpr
  ) {
    super(`(${field.name} ${op} ${other.name})`);
    this.field = field;
    this.op = op;
    this.other = other;
  }

  operator(): Operator {
    return this.op;
  }

  operands(): Expr[] {
    return [this.field, this.other];
  }

  and(that: FieldExpr): BooleanFieldExpr {
    return new BooleanFieldExpr(this, 'and', that);
  }

  or(that: FieldExpr): BooleanFieldExpr {
    return new BooleanFieldExpr(this, 'or', that);
  }

  not(): FieldExpr {
    return new UnaryFieldExpr(this, 'not');
  }
}

export class AggregateFieldExpr<T> extends FieldExpr {
  private field: FieldExpr;
  private op: AggregateOperator;

  constructor(field: FieldExpr, op: AggregateOperator) {
    super(`(${op}(${field.name}))`);
    this.field = field;
    this.op = op;
  }

  override operator(): Operator {
    return this.op;
  }

  override operands(): Expr[] {
    return [this.field];
  }

  override as(alias: string): AggregateFieldExpr<T> {
    super.as(alias);
    return this
  }
}

// TODO(PAT-3177): Define ArithmeticFieldExpr?

File: static/python/0.1.0/src/fieldExpr.py
from typing import Any, Callable, Union, List, Optional, Literal, TypeVar

Scalar = Union[str, int, float, bool]
UnaryOperator = Union[
    Literal["not"],
    Literal["-"],
    Literal["gt"],
    Literal["gte"],
    Literal["lt"],
    Literal["lte"],
]
BooleanOperator = Union[
    Literal["eq"],
    Literal["neq"],
    Literal["gt"],
    Literal["gte"],
    Literal["lt"],
    Literal["lte"],
    Literal["and"],
    Literal["or"],
]
ArithmeticOperator = Union[Literal["+"], Literal["-"], Literal["*"], Literal["/"]]
AggregateOperator = Union[
    Literal["min"], Literal["max"], Literal["count"], Literal["countDistinct"]
]
DateOperator = Union[Literal["years"], Literal["months"], Literal["days"]]
TimeOperator = Union[
    Literal["hour"], Literal["minute"], Literal["second"], Literal["millisecond"]
]
ProjectionOperator = Union[DateOperator, TimeOperator]
DateGranularity = Union[
    Literal["years"], Literal["months"], Literal["weeks"], Literal["days"]
]
TimeGranularity = Union[
    Literal["hours"], Literal["minutes"], Literal["seconds"], Literal["milliseconds"]
]
DateTimeGranularity = Union[DateGranularity, TimeGranularity]
Operator = Union[UnaryOperator, BooleanOperator, ArithmeticOperator, AggregateOperator]


class FieldExpr:

    name: str
    alias: Optional[str] = None

    def __init__(self, name: str, alias: Optional[str] = None):
        self.name = name
        self.alias = alias

    def to_string(self) -> str:
        return self.name

    def as_(self, alias: str) -> "FieldExpr":
        return FieldExpr(self, alias)

    def operator(self) -> Operator:
        pass

    def operands(self) -> List[Union['FieldExpr', Scalar]]:
        return []

Expr = Union[Scalar, FieldExpr]

class UnaryFieldExpr(FieldExpr):
    def __init__(
        self, field: FieldExpr, op: UnaryOperator, alias: Optional[str] = None
    ):
        super().__init__(("(" + op + "(" + field.name + "))"))
        self.field = field
        self.op = op
        self.alias = alias

    def operator(self) -> Operator:
        return self.op

    def operands(self) -> List[Expr]:
        return [self.field]


class BooleanFieldExpr(FieldExpr):
    def __init__(
        self,
        field: FieldExpr,
        op: BooleanOperator,
        other: FieldExpr,
        alias: Optional[str] = None,
    ):
        super().__init__(field, alias)
        self.field = field
        self.op = op
        self.other = other

    def operator(self) -> Operator:
        return self.op

    def operands(self) -> List[Expr]:
        return [self.field, self.other]

    def __and__(self, that: FieldExpr) -> "BooleanFieldExpr":  # &
        return BooleanFieldExpr(self, "and", that)

    def __or__(self, that: FieldExpr) -> "BooleanFieldExpr":  # |
        return BooleanFieldExpr(self, "or", that)

    def not_(self) -> FieldExpr:
        return UnaryFieldExpr(self, "not")


class AggregateFieldExpr(FieldExpr):
    def __init__(self, field: FieldExpr, op: AggregateOperator) -> None:
        super().__init__(f"({op}({field.name}))")
        self.field = field
        self.op = op

    def operator(self) -> Operator:
        return self.op

    def operands(self) -> List[Expr]:
        return [self.field]

    def as_(self, alias: str) -> "AggregateFieldExpr":
        super().as_(alias)
        return self


# TODO(PAT-3177): Define ArithmeticFieldExpr?
